mllib-dal/src/main/native/CorrelationDALImpl.cpp

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include <chrono>
+#include <iostream>
+
+#ifdef CPU_GPU_PROFILE
+#include "GPU.h"
+#endif
+
+#include "OneCCL.h"
+#include "org_apache_spark_ml_stat_CorrelationDALImpl.h"
+#include "service.h"
+
+using namespace std;
+using namespace daal;
+using namespace daal::algorithms;
+
+typedef double algorithmFPType; /* Algorithm floating-point type */
+
+static NumericTablePtr correlation_compute(JNIEnv *env,
+                                           jobject obj,
+                                           int rankId,
+                                           ccl::communicator &comm,
+                                           const NumericTablePtr &pData,
+                                           size_t nBlocks,
+                                           jobject resultObj) {
+
+   auto t1 = std::chrono::high_resolution_clock::now();
+
+   const bool isRoot = (rankId == ccl_root);
+
+   covariance::Distributed<step1Local> localAlgorithm;
+
+  /* Set the input data set to the algorithm */
+   localAlgorithm.input.set(covariance::data, pData);
+
+   /* Compute covariance */
+   localAlgorithm.compute();
+
+   auto t2 = std::chrono::high_resolution_clock::now();
+   auto duration =
+       std::chrono::duration_cast<std::chrono::seconds>(t2 - t1).count();
+   std::cout << "Correleation (native): local step took " << duration << " secs"
+             << std::endl;
+
+   t1 = std::chrono::high_resolution_clock::now();
+
+   /* Serialize partial results required by step 2 */
+   InputDataArchive dataArch;
+   localAlgorithm.getPartialResult()->serialize(dataArch);
+   const uint64_t perNodeArchLength = (size_t)dataArch.getSizeOfArchive();
+
+
+   std::vector<uint64_t> aPerNodeArchLength(comm.size());
+   std::vector<size_t> aReceiveCount(comm.size(), 1);
+   /* Transfer archive length to the step 2 on the root node */
+   ccl::allgatherv(&perNodeArchLength, 1, aPerNodeArchLength.data(), aReceiveCount, comm).wait();
+
+   ByteBuffer serializedData;
+   /* Calculate total archive length */
+   int totalArchLength = 0;
+
+   for (size_t i = 0; i < nBlocks; ++i)
+   {
+       totalArchLength += aPerNodeArchLength[i];
+   }
+   aReceiveCount[ccl_root] = totalArchLength;
+
+   serializedData.resize(totalArchLength);
+
+
+   ByteBuffer nodeResults(perNodeArchLength);
+   dataArch.copyArchiveToArray(&nodeResults[0], perNodeArchLength);
+
+   /* Transfer partial results to step 2 on the root node */
+   ccl::allgatherv((int8_t *)&nodeResults[0], perNodeArchLength, (int8_t *)&serializedData[0], aPerNodeArchLength, comm).wait();
```

Comment:
same as above, you can create ccl::gatherv instead.

Reply:
fixed